### PR TITLE
[FLINK-19272][table-common] Add metadata interfaces for FLIP-107

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

Adds all `@PublicEvolving` interfaces mentioned in FLIP-107.

## Brief change log

- Updated `EncodingFormat` and `DecodingFormat`
- Added `SupportsReadingMetadata` and `SupportsWritingMetadata`

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/abilities/SupportsWritingMetadata.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.utils.LogicalTypeCasts;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Enables to write metadata columns to a {@link DynamicTableSink}.
```
Comment (aljoscha) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/abilities/SupportsWritingMetadata.java:35]: Maybe `Interface for {@link DynamicTableSink DynamicTableSinks} that support writing metadata columns.` sounds better.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/abilities/SupportsWritingMetadata.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+ *
+ * <p>Metadata columns add additional columns to the table's schema. A table sink is responsible to
+ * accept requested metadata columns at the end of consumed rows and persist them. This includes potentially
+ * forwarding metadata columns to contained formats.
```
Comment (aljoscha) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/abilities/SupportsWritingMetadata.java:39]:
```suggestion
 * <p>Metadata columns add additional columns to the table's schema. A table sink is responsible for
 * accepting requested metadata columns at the end of consumed rows and persist them. This includes potentially
 * forwarding metadata columns to contained formats.
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsReadingMetadata.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.utils.LogicalTypeCasts;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Enables to read metadata columns from a {@link ScanTableSource}.
```
Comment (aljoscha) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsReadingMetadata.java:37]: Maybe `Interface for {@link ScanTableSource ScanTableSources} that support reading metadata columns.` sounds better.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsReadingMetadata.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+ *
+ * <p>Metadata columns add additional columns to the table's schema. A table source is responsible to
+ * add requested metadata columns at the end of produced rows. This includes potentially forwarding metadata
+ * columns from contained formats.
```
Comment (aljoscha) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsReadingMetadata.java:41]:
```suggestion
 * <p>Metadata columns add additional columns to the table's schema. A table source is responsible for
 * adding requested metadata columns at the end of produced rows. This includes potentially forwarding metadata
 * columns from contained formats.
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsReadingMetadata.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	 *
+	 * <p>Metadata key names follow the same pattern as mentioned in {@link Factory}. In case of duplicate
+	 * names in format and source keys, format keys shall have higher precedence.
+	 *
+	 * <p>Regardless of the returned {@link DataType}s, a metadata column is always represented using
+	 * internal data structures (see {@link RowData}).
+	 *
+	 * @see DecodingFormat#listReadableMetadata()
+	 */
+	Map<String, DataType> listReadableMetadata();
```
Comment (aljoscha) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsReadingMetadata.java:109]: Returning a `Map` might be a bit dangerous when the order is important. A `List` of tuples would seem better. But I'm sure this was discussed in the FLIP. So please ignore... 😅

**Review by aljoscha — APPROVED** [review]
Approved, but I had some suggestions for wording.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsReadingMetadata.java (hunk 5 above)
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsReadingMetadata.java:109]: The order of the `Map` is not super important. In the end, the passed `List` decides about the order. But an implementation can sync both methods and avoid index mapping if desired.

**Review by aljoscha — APPROVED** [review]

**Pull request closed without merging**